File: problems-adapt/0001-0100/0077_course-feasibility/solution_kahn.js
/**
 * @param {number} courseCount
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var coursesFeasible = function (courseCount, prerequisites) {
    // Each pair [course, prereq] is an edge prereq -> course; all courses can
    // finish exactly when this graph is acyclic.
    const adjacency = Array.from({ length: courseCount }, () => []);
    const indegree = new Array(courseCount).fill(0);
    for (const [course, prereq] of prerequisites) {
        adjacency[prereq].push(course);
        indegree[course] += 1;
    }
    // Kahn's algorithm: seed with every course that has no prerequisites.
    const queue = [];
    for (let i = 0; i < courseCount; i++) {
        if (indegree[i] === 0) {
            queue.push(i);
        }
    }
    let taken = 0;
    let head = 0;
    while (head < queue.length) {
        const node = queue[head++];
        taken += 1;
        // Taking a course removes its outgoing edges.
        for (const nxt of adjacency[node]) {
            indegree[nxt] -= 1;
            if (indegree[nxt] === 0) {
                queue.push(nxt);
            }
        }
    }
    // Courses inside a cycle never reach indegree zero, so a shortfall means
    // a cycle trapped the remainder.
    return taken === courseCount;
};

File: problems/0001-0100/0077_course-feasibility/solution_kahn.go
func coursesFeasible(courseCount int, prerequisites [][]int) bool {
	// Each pair [course, prereq] is an edge prereq -> course; all courses can
	// finish exactly when this graph is acyclic.
	adjacency := make([][]int, courseCount)
	indegree := make([]int, courseCount)
	for _, pair := range prerequisites {
		course, prereq := pair[0], pair[1]
		adjacency[prereq] = append(adjacency[prereq], course)
		indegree[course]++
	}
	// Kahn's algorithm: seed with every course that has no prerequisites.
	queue := make([]int, 0, courseCount)
	for i := 0; i < courseCount; i++ {
		if indegree[i] == 0 {
			queue = append(queue, i)
		}
	}
	taken := 0
	for head := 0; head < len(queue); head++ {
		node := queue[head]
		taken++
		// Taking a course removes its outgoing edges.
		for _, nxt := range adjacency[node] {
			indegree[nxt]--
			if indegree[nxt] == 0 {
				queue = append(queue, nxt)
			}
		}
	}
	// Courses inside a cycle never reach indegree zero, so a shortfall means
	// a cycle trapped the remainder.
	return taken == courseCount
}
